### PR TITLE
PLAT-98205: Separate logics in moonstone/VirtualListBase, ScrollerBase, and ScrollableBase(Native) into custom Hooks

### PR DESCRIPTION
**Pull request by ybsung**
This PR will be merged after the PR - https://github.com/enactjs/enact/pull/2670 is merged and this PR refreshes.

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

There were so many functions defined in moonstone/VirtualListBase, ScrollerBase, and ScrollableBase(Native). It would be better to separate those functions into custom hooks to enhance maintenance and readability.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

Created proper custom hooks.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

There were duplicated logics in moonstone/Scrollable and moonstone/ScrollableNative. By using custom Hooks, I could reduce the duplicated logics. Now two components are quite similar. We could unify them into one in PLAT-98908 easily.

### Links
[//]: # (Related issues, references)

PLAT-98205

### Comments

1. There are useSpottable hooks in moonstone/VirtualList, moonstone/Scrollable, and moonstone/Scroller. We might need to unify similar hooks in those components into one hook in another issue in future.
2. `useCallback` will be applied later in PLAT-98958 because this PR already includes too much.

### CODE REVIEW

File: packages/moonstone/Scrollable/useSpottable.js [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+	const variables = useRef({
+		animateOnFocus: false,
+		indexToFocus: null,
+		lastScrollPositionOnFocus: null,
+		nodeToFocus: null,
+		pointToFocus: null,
+	});
+
+	/*
+	 * useEffects
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useSpottable.js]: Why don't we replace it with a proper name?

File: packages/moonstone/VirtualList/useEventKey.js [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+const useEventKey = (props, instances, dependencies) => {
+	/*
+	 * Dependencies
+	 */
+
+	const {spotlightId} = props;
+	const {spottable: {current: {uiRefCurrent}}} = instances;
+	const {
+		containerNode,
+		handlerGlobalKeyDownCB,
```
Comment (MikyungKim) [inline packages/moonstone/VirtualList/useEventKey.js]: I assume handleGlobalKeyDownCB ?

**Comment by ybsung** [comment]
@MikyungKim 

> Check your indentation option.

Fixed

> maybe... handleGlobalKeyDownCB?

Now removed

> I think we need to import setPointerMode if you want to pass this method. In fact, I'm wondering why we should pass spotlight related methods to useKey.

Passed SpotlightAccelerator, then SpotlightAccelerator.setPointerMode() will be called.

> Please return setContainerDisabled too.
> and also calculatePositionOnFocus.

Fixed

> Why don't we replace it with a proper name?

Fixed  to `* Hooks`

> I assume handleGlobalKeyDownCB ?

Now removed

File: packages/moonstone/Scrollable/useOverscrollEffect.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	 * Dependencies
+	 */
+
+	const {overscrollRefs} = instances;
+
+	/*
+	 * Instance
+	 */
+
+	var variables = useRef({
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useOverscrollEffect.js]: var -> const

File: packages/moonstone/Scrollable/useOverscrollEffect.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+			if (type === overscrollTypeOnce) {
+				variables.current.overscrollJobs[orientation][edge].start(orientation, edge, overscrollTypeDone, 0);
+			}
+		}
+	}
+
+	function clearOverscrollEffect (orientation, edge) {
+		variables.current.overscrollJobs[orientation][edge].startAfter(overscrollTimeout, orientation, edge, overscrollTypeNone, 0);
+		uiRef.current.setOverscrollStatus(orientation, edge, overscrollTypeNone, 0);
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useOverscrollEffect.js]: Where is uiRef?

File: packages/moonstone/Scrollable/useEventWheel.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+		// FIXME This routine is a temporary support for horizontal wheel scroll.
+		// FIXME If web engine supports horizontal wheel, this routine should be refined or removed.
+		if (canScrollVertically) { // This routine handles wheel events on scrollbars for vertical scroll.
+			if (eventDelta < 0 && uiRef.current.scrollTop > 0 || eventDelta > 0 && uiRef.current.scrollTop < bounds.maxTop) {
+				const {horizontalScrollbarRef, verticalScrollbarRef} = uiRef.current;
+
+				if (!variables.current.isWheeling) {
+					if (!props['data-spotlight-container-disabled']) {
+						childRef.current.setContainerDisabled(true);
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useEventWheel.js]: We may remove childRef.current.

File: packages/moonstone/Scrollable/useEventMouse.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		if ((
+			direction === 'vertical' && canScrollVertically(bounds) ||
+			direction === 'horizontal' && canScrollHorizontally(bounds)
+		) && !props['data-spotlight-container-disabled']) {
+			childRef.current.setContainerDisabled(true);
+		}
+	}
+
+	function handleMouseDown (ev) {
+		if (isScrollButtonFocused()) {
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useEventMouse.js]: Can't find this function.

File: packages/moonstone/Scrollable/useEventVoice.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import {useRef} from 'react';
+
+const useEventVoice = (props, instances, dependencies) => {
+	/*
+	 * Dependencies
+	 */
+
+	const {direction} = props;
+	const {uiRef} = instances;
+	const {onScrollbarButtonClick} = dependencies
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useEventVoice.js]: missed ;

File: packages/moonstone/Scrollable/useEventVoice.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+	/*
+	 * Instance
+	 */
+
+	const variables = useRef({
+		isVoiceControl: false,
+		voiceControlDirection: 'vertical',
```
Comment (MikyungKim) [inline packages/moonstone/Scrollable/useEventVoice.js]: additional ,

File: packages/moonstone/Scroller/useSpotlightConfig.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	useEffect(() => {
+		configureSpotlight();
+	}, [spotlightId]);	// TODO : Handle exhaustive-deps ESLint rule.
+
+	/*
+	 * Functions
+	 */
+
+	function handleLeaveContainer ({direction, target}) {
+		const {current: {containerRef, scrollBounds: {maxLeft, maxTop}, scrollPos: {left, top}}} = uiRef.current;
```
Comment (MikyungKim) [inline packages/moonstone/Scroller/useSpotlightConfig.js]: suspicious current: part.

File: packages/moonstone/Scroller/useSpottable.js [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+	 * Hooks
+	 */
+
+	useSpotlightConfig(instances, props);
+
+	const {addGlobalKeyDownEventListener, removeGlobalKeyDownEventListener} = useEventKey();
+
+	useEffect(() => {
+		return () => setContainerDisabled(false);
+	}, [setContainerDisabled]);	// TODO : Handle exhaustive-deps ESLint rule.
```
Comment (MikyungKim) [inline packages/moonstone/Scroller/useSpottable.js]: I think it depends on `spotlightId`, not `setContainerDisabled`.

**Comment by 0x64** [comment]
LGTM!

**Pull request merged**